File: apps/api/pkg/domain/route_test.go
package domain

import "testing"

// The index, reduced to the rows these tests need. The real lookup comes from
// pkg/services/airports.
var testAirports = map[string]Airport{
	"BKK": {IATA: "BKK", City: "Bangkok", CityTH: "กรุงเทพ", CountryCode: "TH", Country: "Thailand", CountryTH: "ไทย"},
	"NRT": {IATA: "NRT", City: "Tokyo", CityTH: "โตเกียว", CountryCode: "JP", Country: "Japan", CountryTH: "ญี่ปุ่น"},
	"KIX": {IATA: "KIX", City: "Osaka", CityTH: "โอซาก้า", CountryCode: "JP", Country: "Japan", CountryTH: "ญี่ปุ่น"},
	"ICN": {IATA: "ICN", City: "Seoul", CityTH: "โซล", CountryCode: "KR", Country: "South Korea", CountryTH: "เกาหลีใต้"},
}

func lookup(iata string) *Airport {
	if a, ok := testAirports[iata]; ok {
		return &a
	}
	return nil
}

// The trip from the brief: BKK→NRT arriving 08:05 on 4 Dec, NRT→BKK arriving
// 22:05 on 10 Dec.
func TestBuildRouteReturnTrip(t *testing.T) {
	route := BuildRoute([]Leg{
		{Direction: "out", DepAirport: "BKK", ArrAirport: "NRT", DepDate: "2026-12-04", ArrDate: "2026-12-04", ArrTime: "08:05"},
		{Direction: "back", DepAirport: "NRT", ArrAirport: "BKK", DepDate: "2026-12-10", ArrDate: "2026-12-10", ArrTime: "22:05"},
	}, lookup)

	if !route.RoundTrip {
		t.Error("landing back at BKK should read as a round trip")
	}
	if route.StartDate != "2026-12-04" || route.EndDate != "2026-12-10" {
		t.Errorf("frame is %s → %s", route.StartDate, route.EndDate)
	}
	if route.Days != 7 || route.Nights != 6 {
		t.Errorf("got %d days / %d nights, want 7/6", route.Days, route.Nights)
	}
	if len(route.Stops) != 1 {
		t.Fatalf("got %d stops, want only Tokyo", len(route.Stops))
	}

	stop := route.Stops[0]
	if stop.City != "โตเกียว" || stop.Nights != 6 || stop.ArriveTime != "08:05" || stop.Open {
		t.Errorf("unexpected stop: %+v", stop)
	}
	if got := route.Cities(); len(got) != 1 || got[0] != "โตเกียว" {
		t.Errorf("cities = %v", got)
	}
	if route.PrimaryCountry() != "JP" {
		t.Errorf("primary country = %q", route.PrimaryCountry())
	}
}

// "Seoul and Tokyo" — the ambiguity that used to be a chip list. Two countries,
// and the nights land where the legs say they land.
func TestBuildRouteTwoCountries(t *testing.T) {
	route := BuildRoute([]Leg{
		{Direction: "out", DepAirport: "BKK", ArrAirport: "ICN", DepDate: "2026-12-04", ArrTime: "07:30"},
		{Direction: "inter", DepAirport: "ICN", ArrAirport: "NRT", DepDate: "2026-12-07", ArrTime: "12:10"},
		{Direction: "back", DepAirport: "NRT", ArrAirport: "BKK", DepDate: "2026-12-11", ArrTime: "22:05"},
	}, lookup)

	if len(route.Stops) != 2 {
		t.Fatalf("got %d stops, want Seoul then Tokyo", len(route.Stops))
	}
	if route.Stops[0].City != "โซล" || route.Stops[0].Nights != 3 {
		t.Errorf("first stop: %+v", route.Stops[0])
	}
	if route.Stops[1].City != "โตเกียว" || route.Stops[1].Nights != 4 {
		t.Errorf("second stop: %+v", route.Stops[1])
	}

	if len(route.Countries) != 2 {
		t.Fatalf("got %d countries, want 2: %+v", len(route.Countries), route.Countries)
	}
	if route.Countries[0].Code != "KR" || route.Countries[0].Nights != 3 {
		t.Errorf("first country: %+v", route.Countries[0])
	}
	if route.Countries[1].Code != "JP" || route.Countries[1].Nights != 4 {
		t.Errorf("second country: %+v", route.Countries[1])
	}
	if route.PrimaryCountry() != "JP" {
		t.Errorf("primary country = %q, want the one with most nights", route.PrimaryCountry())
	}
}

// Two cities in one country still aggregate into one country line.
func TestBuildRouteTwoCitiesOneCountry(t *testing.T) {
	route := BuildRoute([]Leg{
		{DepAirport: "BKK", ArrAirport: "NRT", DepDate: "2026-12-04"},
		{Mode: "ground", DepAirport: "NRT", ArrAirport: "KIX", DepDate: "2026-12-08"},
		{DepAirport: "KIX", ArrAirport: "BKK", DepDate: "2026-12-11"},
	}, lookup)

	if len(route.Countries) != 1 || route.Countries[0].Code != "JP" || route.Countries[0].Nights != 7 {
		t.Fatalf("countries = %+v", route.Countries)
	}
	if route.Countries[0].Cities != "โตเกียว · โอซาก้า" {
		t.Errorf("cities line = %q", route.Countries[0].Cities)
	}
}

// A one-way ticket is a real answer: the last stop simply stays open.
func TestBuildRouteOneWay(t *testing.T) {
	route := BuildRoute([]Leg{
		{DepAirport: "BKK", ArrAirport: "NRT", DepDate: "2026-12-04", ArrTime: "08:05"},
	}, lookup)

	if route.RoundTrip {
		t.Error("one leg is not a round trip")
	}
	if len(route.Stops) != 1 || !route.Stops[0].Open {
		t.Fatalf("stops = %+v", route.Stops)
	}
	if route.Stops[0].Nights != 0 {
		t.Errorf("an open stop has no night count yet, got %d", route.Stops[0].Nights)
	}
}

// Legs arrive in whatever order the user typed them; the route is chronological.
func TestBuildRouteSortsAndSkipsIncompleteLegs(t *testing.T) {
	route := BuildRoute([]Leg{
		{DepAirport: "NRT", ArrAirport: "BKK", DepDate: "2026-12-10"},
		{DepAirport: "BKK", ArrAirport: "NRT", DepDate: "2026-12-04"},
		{DepAirport: "BKK", ArrAirport: "", DepDate: "2026-12-04"}, // half-typed
		{DepAirport: "BKK", ArrAirport: "KIX"},                     // no date yet
	}, lookup)

	if route.StartDate != "2026-12-04" || route.EndDate != "2026-12-10" {
		t.Errorf("frame is %s → %s", route.StartDate, route.EndDate)
	}
	if len(route.Stops) != 1 || route.Stops[0].Airport != "NRT" {
		t.Errorf("stops = %+v", route.Stops)
	}
}

// An airport the index has never heard of must not disappear from the route.
func TestBuildRouteUnknownAirport(t *testing.T) {
	route := BuildRoute([]Leg{
		{DepAirport: "BKK", ArrAirport: "ZZZ", DepDate: "2026-12-04"},
		{DepAirport: "ZZZ", ArrAirport: "BKK", DepDate: "2026-12-08"},
	}, lookup)

	if len(route.Stops) != 1 || route.Stops[0].City != "ZZZ" || route.Stops[0].Nights != 4 {
		t.Fatalf("stops = %+v", route.Stops)
	}
}

func TestBuildRouteEmpty(t *testing.T) {
	route := BuildRoute(nil, lookup)
	if len(route.Stops) != 0 || route.Days != 0 || route.StartDate != "" {
		t.Errorf("empty route should stay empty: %+v", route)
	}
}

File: apps/web/lib/__tests__/route.test.ts
import { describe, expect, it } from 'vitest';

import { searchAirports } from '@/lib/data/airports';
import { buildRoute, routeCities, routeWarnings, stayLabel } from '@/lib/data/route';
import type { Airport, FlightLegInput } from '@/lib/data';

/**
 * The twin of `apps/api/pkg/domain/route_test.go` — same legs, same numbers.
 * Mock mode derives the route in the browser and live mode receives it from the
 * API, and a UAT tester must not be able to tell which one drew the screen.
 */

function airport(iata: string, cityTh: string, cc: string, countryTh: string): Airport {
  return {
    iata,
    name: `${cityTh} Airport`,
    city: cityTh,
    cityTh,
    countryCode: cc,
    country: cc,
    countryTh,
    timezone: 'Asia/Bangkok',
    lat: 0,
    lon: 0,
    major: true,
  };
}

const INDEX: Record<string, Airport> = {
  BKK: airport('BKK', 'กรุงเทพ', 'TH', 'ไทย'),
  NRT: airport('NRT', 'โตเกียว', 'JP', 'ญี่ปุ่น'),
  KIX: airport('KIX', 'โอซาก้า', 'JP', 'ญี่ปุ่น'),
  ICN: airport('ICN', 'โซล', 'KR', 'เกาหลีใต้'),
};

const lookup = (iata: string) => INDEX[iata] ?? null;

function leg(patch: Partial<FlightLegInput>): FlightLegInput {
  return { direction: 'out', mode: 'flight', from: '', to: '', depDate: '', ...patch };
}

describe('buildRoute', () => {
  // The trip from the brief: BKK→NRT 4 ธ.ค. ถึง 08:05, NRT→BKK 10 ธ.ค. ถึง 22:05.
  it('turns a return ticket into a frame', () => {
    const route = buildRoute(
      [
        leg({ from: 'BKK', to: 'NRT', depDate: '2026-12-04', arrTime: '08:05' }),
        leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-10', arrTime: '22:05' }),
      ],
      lookup,
    );

    expect(route.roundTrip).toBe(true);
    expect(route.startDate).toBe('2026-12-04');
    expect(route.endDate).toBe('2026-12-10');
    expect(route.days).toBe(7);
    expect(route.nights).toBe(6);
    expect(route.stops).toHaveLength(1);
    expect(route.stops[0]).toMatchObject({ city: 'โตเกียว', nights: 6, arriveTime: '08:05' });
    expect(routeCities(route)).toEqual(['โตเกียว']);
  });

  // "โซล อูเอโนะ" used to be unanswerable. As legs it is two countries with a
  // night count each.
  it('splits a two-country route by nights', () => {
    const route = buildRoute(
      [
        leg({ from: 'BKK', to: 'ICN', depDate: '2026-12-04' }),
        leg({ direction: 'inter', from: 'ICN', to: 'NRT', depDate: '2026-12-07' }),
        leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-11' }),
      ],
      lookup,
    );

    expect(route.countries).toEqual([
      { code: 'KR', name: 'เกาหลีใต้', cities: 'โซล', nights: 3 },
      { code: 'JP', name: 'ญี่ปุ่น', cities: 'โตเกียว', nights: 4 },
    ]);
    expect(stayLabel(route)).toBe('โซล 3 คืน · โตเกียว 4 คืน');
  });

  it('keeps two cities of one country on one country line', () => {
    const route = buildRoute(
      [
        leg({ from: 'BKK', to: 'NRT', depDate: '2026-12-04' }),
        leg({ direction: 'inter', mode: 'ground', from: 'NRT', to: 'KIX', depDate: '2026-12-08' }),
        leg({ direction: 'back', from: 'KIX', to: 'BKK', depDate: '2026-12-11' }),
      ],
      lookup,
    );

    expect(route.countries).toHaveLength(1);
    expect(route.countries[0]).toMatchObject({ code: 'JP', nights: 7, cities: 'โตเกียว · โอซาก้า' });
  });

  it('leaves the last stop open on a one-way ticket', () => {
    const route = buildRoute([leg({ from: 'BKK', to: 'NRT', depDate: '2026-12-04' })], lookup);

    expect(route.roundTrip).toBe(false);
    expect(route.stops[0]?.open).toBe(true);
    expect(route.stops[0]?.nights).toBe(0);
  });

  it('sorts by date and ignores half-typed legs', () => {
    const route = buildRoute(
      [
        leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-10' }),
        leg({ from: 'BKK', to: 'NRT', depDate: '2026-12-04' }),
        leg({ from: 'BKK', to: '', depDate: '2026-12-04' }),
        leg({ from: 'BKK', to: 'KIX' }),
      ],
      lookup,
    );

    expect(route.startDate).toBe('2026-12-04');
    expect(route.stops.map((s) => s.airport)).toEqual(['NRT']);
  });

  it('counts an overnight flight against the arrival date', () => {
    const route = buildRoute(
      [
        leg({ from: 'BKK', to: 'NRT', depDate: '2026-12-03', depTime: '23:59', arrDate: '2026-12-04', arrTime: '08:05' }),
        leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-10' }),
      ],
      lookup,
    );

    expect(route.stops[0]).toMatchObject({ arriveDate: '2026-12-04', nights: 6 });
  });

  it('keeps an airport the index has never heard of', () => {
    const route = buildRoute(
      [
        leg({ from: 'BKK', to: 'ZZZ', depDate: '2026-12-04' }),
        leg({ direction: 'back', from: 'ZZZ', to: 'BKK', depDate: '2026-12-08' }),
      ],
      lookup,
    );

    expect(route.stops[0]).toMatchObject({ city: 'ZZZ', nights: 4 });
  });

  it('is empty for no legs', () => {
    const route = buildRoute([], lookup);
    expect(route.stops).toHaveLength(0);
    expect(route.days).toBe(0);
  });
});

describe('routeWarnings', () => {
  it('names the gap when the next leg leaves from somewhere else', () => {
    const legs = [
      leg({ from: 'BKK', to: 'ICN', depDate: '2026-12-04' }),
      leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-10' }),
    ];
    const warnings = routeWarnings(buildRoute(legs, lookup), legs);

    expect(warnings.some((w) => w.id === 'gap-ICN-NRT' && w.level === 'warn')).toBe(true);
  });

  it('says how many countries and how many nights each gets', () => {
    const legs = [
      leg({ from: 'BKK', to: 'ICN', depDate: '2026-12-04' }),
      leg({ direction: 'inter', from: 'ICN', to: 'NRT', depDate: '2026-12-07' }),
      leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-11' }),
    ];
    const warnings = routeWarnings(buildRoute(legs, lookup), legs);
    const multi = warnings.find((w) => w.id === 'multi-country');

    expect(multi?.text).toContain('2 ประเทศ');
    expect(multi?.text).toContain('เกาหลีใต้ 3 คืน');
  });

  it('flags a stop that gets one night', () => {
    const legs = [
      leg({ from: 'BKK', to: 'ICN', depDate: '2026-12-04' }),
      leg({ direction: 'inter', from: 'ICN', to: 'NRT', depDate: '2026-12-05' }),
      leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-11' }),
    ];
    const warnings = routeWarnings(buildRoute(legs, lookup), legs);

    expect(warnings.some((w) => w.id === 'short-ICN')).toBe(true);
  });

  it('says nothing about a clean return trip', () => {
    const legs = [
      leg({ from: 'BKK', to: 'NRT', depDate: '2026-12-04' }),
      leg({ direction: 'back', from: 'NRT', to: 'BKK', depDate: '2026-12-10' }),
    ];

    expect(routeWarnings(buildRoute(legs, lookup), legs)).toEqual([]);
  });
});

/**
 * The index itself. It is the same JSON the Go service embeds, so these are the
 * searches a Thai group actually types.
 */
describe('searchAirports', () => {
  it('puts the exact IATA code first', async () => {
    const found = await searchAirports('NRT', 5);
    expect(found[0]?.iata).toBe('NRT');
    expect(found[0]?.cityTh).toBe('โตเกียว');
    expect(found[0]?.countryTh).toBe('ญี่ปุ่น');
  });

  it('searches in Thai', async () => {
    const found = await searchAirports('โตเกียว', 5);
    expect(found.map((a) => a.iata)).toContain('HND');
  });

  it('reaches every continent, not only the ones we seeded', async () => {
    for (const [query, iata] of [
      ['queenstown', 'ZQN'],
      ['cape town', 'CPT'],
      ['lisbon', 'LIS'],
      ['reykjavik', 'RKV'],
    ] as const) {
      const found = await searchAirports(query, 8);
      expect(found.map((a) => a.iata)).toContain(iata);
    }
  });

  it('answers an empty query with the hubs', async () => {
    const found = await searchAirports('', 6);
    expect(found).toHaveLength(6);
    expect(found[0]?.iata).toBe('BKK');
  });
});
